Source/Controllers/Other/SecureBookmarkManager.swift

```diff
@@ -0,0 +1,336 @@
+//
+//  SecureBookmarkManager.swift
+//  Sequel Ace
+//
+//  Created by James on 7/12/2020.
+//  Copyright © 2020 Sequel-Ace. All rights reserved.
+//
+
+import Foundation
+import os.log
+import Firebase
+
+
+/*
+reRequestSecureAccess
+addBookmark
+handle bookmarkDataIsStale
+revokeBookmark
+stopAllSecurityScopedAccess
+
+*/
+
+@objc final class SecureBookmarkManager: NSObject {
+
+	@objc static let sharedInstance = SecureBookmarkManager()
+
+    @objc public var bookmarks: [Dictionary<String, Data>] = []
+	@objc public var resolvedBookmarks: [URL] = []
+	@objc public var staleBookmarks: [String] = []
+
+    private let _NSURLBookmarkResolutionWithSecurityScope = URL.BookmarkResolutionOptions(rawValue: 1 << 10)
+    private let log: OSLog
+    private let prefs: UserDefaults = UserDefaults.standard
+    private var observer: NSKeyValueObservation?
+
+    private var iChangedTheBookmarks: Bool = false
+
+
+	override private init() {
+		log = OSLog(subsystem: Bundle.main.bundleIdentifier!, category: "secureBookmarks")
```

Comment:
Try to avoid `!`

Reply:
I am ... very confused by the optionals

Reply:
Why is that? :)

Reply:
?? ... ? .... ! ?as..... as?  as!  try?  try! ahhhhh

Reply:
It depends - you can make the `log` itself optional, which means to this in `if let bundleID = Bundle.main.bundleIdentifier` - if it's just a logger that we are okay with not exist in worst case.

You can also guard the bundle ID, if you think the app should fail / show error and not continue if the ID doesn't exist. :)

Reply:
`try` (and it's variations) are for methods that can `throw` (in objective C - methods that accept `&error` - instead of accepting the error variable swift will either jump to `catch` block or return nil if you do `try?` - you are saying compiler try to get the value, if not, give me optional)

`as?` is usually for things that have some type and you want subtype, or values in dictionary / array and you want their specific type

`!` (in any variation) shouldn't be used at all - it's workaround around compiler type safety checks that shouldn't be done if you don't want to encounter crashes.

`??` is the same as objective `?:` - like `value != nil ?: otherValue` - so in swift you do `value ?? other value`

Reply:
thanks for this.